flink-job/src/main/java/com/ververica/field/dynamicrules/functions/TimeStamper.java

```diff
@@ -25,7 +25,7 @@
 public class TimeStamper<T extends TimestampAssignable<Long>> extends RichFlatMapFunction<T, T> {
 
   @Override
-  public void flatMap(T value, Collector<T> out) throws Exception {
+  public void flatMap(T value, Collector<T> out) {
```

Comment:
I am not sure about this one. I usually like to preserve the full signature of the method that is being overridden. There is also some special handling of the exception by the framework that is underlined in the parent. Leaving it in the signature communicates that there is something to take into account here:
```
flatMap in RichFlatMapFunction:

@throws Exception This method may throw exceptions. Throwing an exception will cause the operation to fail and may trigger recovery.
```

Reply:
There is nothing to take into account here though, i.e. for this function. Since it doesn't throw, it should also not advertise that (unless it is meant to be overridden again and this is an API construct; this doesn't really apply here, does it?)

Reply:
I have a couple of comments:
1. This is what we do in all of our training exercises -> consistency of externally-communicated material should be something we should strive for:
https://github.com/ververica/flink-training-exercises/blob/179f940987dcf643f629051af22ce4ad67a8aa66/src/main/java/com/ververica/flinktraining/solutions/datastream_java/process/LongRidesSolution.java
2. I see no harm in being non-restrictive regarding inheritance in such public "exercises". I can see people wanting to extend our classes. By removing the throws we limit them.
3. This "throws" is in there for a reason in the framework, and this reason is documented in the comment. Leaving it reveals some important bit of information to the new users.
4. This is the default behavior of IntelliJ if you use the "Override ", and i see no good reason to fight it.

Reply:
If those arguments are not convincing, feel free to remove it, I am fine with that.

Reply:
Although this change isn't too important (as it doesn't change much, code-wise), I don't find your arguments convincing and would keep it.
I find it important (and actually necessary) for an API or library, but not for a concrete implementation (with the only limitation that it reduces extensibility which we do not require here).

1. actually, the new training exercises are on https://github.com/apache/flink-training and I got this change through this code as well, see https://github.com/apache/flink-training/blob/6f9ddff6e9da835ef2f526113d8a5fa98998466c/long-ride-alerts/src/solution/java/org/apache/flink/training/solutions/longrides/LongRidesSolution.java
2. we don't even publish artifacts...if people want to extend this code, they'd have to fork it anyway, making all changed they want
3. I disagree: we're not making use of that possibility from the API. It **allows throwing** an exception but this class **does not throw** any (checked) exception!
4. I'm only assuming here that you are talking about the skeleton implementation that you get when asking IntelliJ to create an override...this doesn't even have any code yet and therefore keeps everything. It does however come with a "Redundant 'throws' clause" inspection telling you this:
> Exception 'java.lang.Exception' is never thrown in the method 
 Inspection info: Reports exceptions that are declared in a method's signature but never thrown by the method itself or its implementations/derivatives.

Reply:
What I see in the link you've shared is this:
```public void processElement(TaxiRide ride, Context context, Collector<TaxiRide> out) throws Exception {```

Reply:
oh, yes, sorry - that one also actually may throw an exception during state updates (bad example)...this one is better: https://github.com/apache/flink-training/blob/6f9ddff6e9da835ef2f526113d8a5fa98998466c/ride-cleansing/src/solution/java/org/apache/flink/training/solutions/ridecleansing/RideCleansingSolution.java#L77